overlord/snapstate/snapmgr.go

```diff
@@ -551,6 +556,107 @@ func (m *SnapManager) ensureUbuntuCoreTransition() error {
+// profiles for the snap-confine binary from core
+func (m *SnapManager) cleanupSnapConfineApparmor(currentSnapConfineProfilePath string) error {
+	apparmorProfilePathPattern := strings.Replace(filepath.Join(dirs.SnapMountDir, "/core/*/usr/lib/snapd/snap-confine"), "/", ".", -1)[1:]
+
+	glob, err := filepath.Glob(filepath.Join(dirs.SystemApparmorDir, apparmorProfilePathPattern))
+	if err != nil {
+		return err
+	}
+
+	for _, path := range glob {
+		if path == currentSnapConfineProfilePath {
+			continue
+		}
+
+		// not using apparmor.UnloadProfile() because it uses a
+		// different cachedir
+		if output, err := exec.Command("apparmor_parser", "-R", filepath.Base(path)).CombinedOutput(); err != nil {
+			logger.Noticef("cannot unload apparmor profile %s: %v", filepath.Base(path), osutil.OutputErr(output, err))
+		}
+		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		if err := os.Remove(filepath.Join(dirs.SystemApparmorCacheDir, filepath.Base(path))); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *SnapManager) addSnapConfineApparmor(snapConfineInCore, apparmorProfilePath string) error {
+	coreRoot := filepath.Join(dirs.SnapMountDir, "/core/current/")
+
+	// FIXME: make this more generic once we start supporting more
+	//        distros with apparmor around snap-confine, i.e.
+	//        ship `snap-confine.apparmor.in` and use that as the base
+	//        because core is ubuntu and classic host might be anything
+	apparmorProfile, err := ioutil.ReadFile(filepath.Join(coreRoot, "/etc/apparmor.d/usr.lib.snapd.snap-confine"))
+	if err != nil {
+		return err
+	}
+
+	apparmorProfileForCore := strings.Replace(string(apparmorProfile), "/usr/lib/snapd/snap-confine", snapConfineInCore, -1)
+
+	// /etc/apparmor.d is read/write OnClassic, so write out the
+	// new core's profile there
+	if err := osutil.AtomicWriteFile(apparmorProfilePath, []byte(apparmorProfileForCore), 0644, 0); err != nil {
+		return err
+	}
+
+	// not using apparmor.LoadProfile() because it uses a different cachedir
+	if output, err := exec.Command("apparmor_parser", "--replace", "--write-cache", apparmorProfilePath, "--cache-loc", dirs.SystemApparmorCacheDir).CombinedOutput(); err != nil {
+		return fmt.Errorf("cannot replace snap-confine apparmor profile: %v", osutil.OutputErr(output, err))
+	}
+
+	return nil
+}
+
+// ensureSnapConfineApparmor ensures that we have a valid apparmor
+// profile for snap-confine when we are in re-exec mode
+func (m *SnapManager) ensureSnapConfineApparmor() error {
+	// On all-snaps we always use the mounted snapd, snap-confine,
+	// apparmor, etc but on classic we use the snapd and
+	// snap-confine from the latest core snap.  As such, nothing
+	// to do when not OnClassic
+	if !release.OnClassic {
+		return nil
+	}
+	// On releases that do not support apparmor there is no need to
+	// write an apparmor profile for snap-confine
+	if release.ReleaseInfo.ForceDevMode() {
```

Comment:
That is not true. What will happen is that snap-confine from core will try to load one anyway (because today that is a compile time choice) and it will abort because there's no profile to load.

I've started teaching snap-confine to do runtime detection but it won't be ready for this week.